### ISSUE TITLE
Question: Steps necessary to POST a registration using REST API?

### ISSUE THREAD

**Issue by jasonnehf** [eventespresso/event-espresso-core#283]
### Is this a bug report?

- [ ] yes
- [x] no

Could someone please list the steps necessary to register a person to an event using the REST API? 

I've been banging my head against it for the past two days, and every time I think I'm making some headway, I meet some gotcha or some issue that causes that registration to misbehave.

So far, here's what I've got:
(pseudocode, of course)
```
for i in incoming_registrations
	parse unique event_code from i
	GET event from /events?(where EVT_name LIKE %evt_code%), store response as event
	parse p.primary for attendee info, add to attendees
	parse p.guests for attendee info, add to attendees
	for a in attendees
		POST a.attendee_info to /attendees?, store response as att
		POST new transaction to /transactions?, store response as txn
		POST event.EVT_ID, att.ATT_ID, txn.TXN_ID, etc. to /registrations
```
Which _almost_ works! But when I try to change the status of the new registration via the EE Registrations page (I have them initially 'not-accepted', since they're coming from an external source), the page errors out because it can't call ticket-related functions on null obejcts. Because the registration's ticket wasn't ever created. Well **_cripes_**, it would sure help to know what else has to be done for this to all work properly?

Am I almost there? What else has to be done here? The docs for the Write portion of the REST API are woefully inadequate when it comes to specifics. They tell you how to make a general POST, but not how to accomplish a real-life task (like registering a new attendee). Hence all the head-banging.

Any help? 

(I'm _secretly_ hoping I've been doing this all wrong and there's a single endpoint that I can POST all this data to and it takes care of it for me. But it's never that easy, is it?)

**Comment by tn3rb** [comment]
Are you simply trying to create a single record in the `wp_esp_registration` table or are you trying to simulate **everything** that happens when a registration is created normally using the registration checkout form on the frontend of a site?

If you are trying to do the latter, then you absolutely need to create and/or add relations to **ALL** of the objects that are associated with a registration, such as : 

 * Transaction
 * Event
 * Attendee
 * Ticket

> the page errors out because it can't call ticket-related functions on null obejcts.

right, because our system expects a registration to have been created as a result of someone selecting a ticket for an event. If a registration does not have an associated ticket... then what is it registering for?

Here's all of the fields on the `wp_esp_registration` table where a non-null value is required:

 * REG_ID
 * EVT_ID
 * ATT_ID
 * TXN_ID
 * TKT_ID
 * STS_ID (defaults to: RPP)
 * REG_date (defaults to: 0000-00-00 00:00:00)
 * REG_final_price (defaults to: 0.000)
 * REG_paid (defaults to: 0.000)
 * REG_session

although a `REG_code` and `REG_url_link` should also be added for things to work correctly but are technically not required when a record is first created.

@mnelson4 (our API's primary engineer) may have some additional insights to help you out

**Comment by jasonnehf** [comment]
Here's all of the fields on the wp_esp_registration table where a non-null value is required: 

- REG_ID
- EVT_ID
- ATT_ID
- TXN_ID
- TKT_ID
- STS_ID (defaults to: RPP)
- REG_date (defaults to: 0000-00-00 00:00:00)
- REG_final_price (defaults to: 0.000)
- REG_paid (defaults to: 0.000)
- REG_session
- 
-

**Comment by jasonnehf** [comment]
I should mention that the service I'm trying to implement is meant to bridge two registration systems together -- the client has a separate phone service that registers to their internal system, and they want those phone-based registrations added to the EE system (and EE-based registrations to their phone system, but that's not your problem). So here I am, trying to bridge two systems.

> Here's all of the fields on the wp_esp_registration table where a non-null value is required:
> 
> - REG_ID
> - EVT_ID
> - ATT_ID
> - TXN_ID
> - TKT_ID
> - STS_ID (defaults to: RPP)
> - REG_date (defaults to: 0000-00-00 00:00:00)
> - REG_final_price (defaults to: 0.000)
> - REG_paid (defaults to: 0.000)
> - REG_session
> - REG_code
> - REG_url_link

> a non-null value is required

Why aren't those fields specified as 'required' in the API? 

I mean, I recognize that, yes, one can create a registration without many (any?) of these values specified, but the resulting registration will be broken in almost every measurable way. 

Regardless. Okay -- so it looks like I'll need to create a ticket. But I'm looking at the /tickets? endpoint, and I wilt -- **_I wilt !!!_** -- when I see everything that it's needs to be valid. Looks like I can either fill in every single field myself (nope, since my system has no knowledge of these values) or use a ticket_template. I hoped that the ticket_template would be attached to the event, but I don't see one. How am I to decide which ticket_template to use? Or, barring that, how is an external system supposed to know the details of a ticket for a registration of an attendee of an event? (Note to @mnelson4 : event registration may be a candidate for a new RPC-style endpoint, because it's _exhausting_ to do by hand.)

As for REG_session, REG_code, and REG_url_link, I'm making these calls externally, so I'm not sure what the REG_session could be referring to. Can I enter any GUID in there? Similar question for the REG_code -- I have no idea how that value is normally generated. (I imagine I could use relatively simple string concatenation to generate REG_url_link.)

I'm strongly considering writing all this up as a how-to doc once this is all over, since I imagine I'm not the first and won't be the last to perform this (grueling) task. :)

**Comment by tn3rb** [comment]
> the client has a separate phone service that registers to their internal system, and they want those phone-based registrations added to the EE system (and EE-based registrations to their phone system...

ya trying to get apples and bananas to function like they are the same thing doesn't sound like a lot of fun to me

> Why aren't those fields specified as 'required' in the API?

because they are not required for the API requests to be successful which is not the same thing as having a complete data model that has been fully populated with its required data. Think about web applications where a user can edit single fields which get updated immediately on the server. You are not **required** to send every single field with each request, but you are still expected to fully complete all of the data necessary for that data model to be completed. So you can send a request to update a single property but you can not leave data objects half populated and expect the system to still work properly. Make sense?

> I'm looking at the /tickets? endpoint, and I wilt -- I wilt !!! -- when I see everything that it's needs to be valid.

dunno what to say other than...  yup! Our data model is pretty complex because our software has extensive capabilities. 

>  I hoped that the ticket_template would be attached to the event, but I don't see one. How am I to decide which ticket_template to use? 

Are you referring to the "TTM_ID" or "Ticket Template ID" field? 
That's for specifying an actual HTML template to be used when printing physical tickets. 
Just use the default value for that.

> event registration may be a candidate for a new RPC-style endpoint, because it's exhausting to do by hand.

Our REST API is still relatively new but we eventually plan to add endpoints that are for more than just direct i/o type manipulation of the data models.

> As for REG_session, REG_code, and REG_url_link, I'm making these calls externally, so I'm not sure what the REG_session could be referring to. Can I enter any GUID in there? Similar question for the REG_code -- I have no idea how that value is normally generated.

I highly recommend looking at our underlying data models which can be found in:

```
/core/db_models/
```

as well as the resulting data objects that get constructed, which can be found in ::

```
/core/db_classes/
```

REG_session is just a identifier from the user session when the registration was created. You can likely use any similar UUID from your other system.
Reg_url_link and Reg_code are both ultimately created by Value Objects which can be found at:

```
\core\domain\entities\RegUrlLink
\core\domain\entities\RegCode
```

but looking at t those classes will give you an idea of how those strings are generated.

> I'm strongly considering writing all this up as a how-to doc once this is all over, since I imagine I'm not the first and won't be the last to perform this task. :)

That would be great. We would love to see that. 
Active contributors in our community that are professional developers can also get referral business from us: https://eventespresso.com/developers/event-espresso-pros/
Don't think we have anyone on there that specializes in the API.

**Comment by nerrad** [comment]
Another thing you can do is make a `OPTIONS` request to any endpoint and it will return you a schema of that endpoint which includes which relations exist for that endpoint.  That can help guide what relations apply to a registration.  For instance if you do a `OPTIONS` request to `/wp-json/ee/v4.8.36/registrations` - you'll discover this subset of json objects in the `schema` property of the response:

```
"event": {
        "description": "The related Event entity to the Registration.",
        "type": "object",
        "relation": true,
        "relation_type": "EE_Belongs_To_Relation",
        "readonly": true,
        "relation_model": "Event"
      },
      "attendee": {
        "description": "The related Attendee entity to the Registration.",
        "type": "object",
        "relation": true,
        "relation_type": "EE_Belongs_To_Relation",
        "readonly": true,
        "relation_model": "Attendee"
      },
      "transaction": {
        "description": "The related Transaction entity to the Registration.",
        "type": "object",
        "relation": true,
        "relation_type": "EE_Belongs_To_Relation",
        "readonly": true,
        "relation_model": "Transaction"
      },
      "ticket": {
        "description": "The related Ticket entity to the Registration.",
        "type": "object",
        "relation": true,
        "relation_type": "EE_Belongs_To_Relation",
        "readonly": true,
        "relation_model": "Ticket"
      },
      "status": {
        "description": "The related Status entity to the Registration.",
        "type": "object",
        "relation": true,
        "relation_type": "EE_Belongs_To_Relation",
        "readonly": true,
        "relation_model": "Status"
      },
      "answers": {
        "description": "The related Answer entities to the Registration.",
        "type": "array",
        "relation": true,
        "relation_type": "EE_Has_Many_Relation",
        "readonly": true,
        "items": {
          "type": "object"
        },
        "relation_model": "Answer"
      },
      "checkins": {
        "description": "The related Checkin entities to the Registration.",
        "type": "array",
        "relation": true,
        "relation_type": "EE_Has_Many_Relation",
        "readonly": true,
        "items": {
          "type": "object"
        },
        "relation_model": "Checkin"
      },
      "registration_payments": {
        "description": "The related Registration_Payment entities to the Registration.",
        "type": "array",
        "relation": true,
        "relation_type": "EE_Has_Many_Relation",
        "readonly": true,
        "items": {
          "type": "object"
        },
        "relation_model": "Registration_Payment"
      },
      "payments": {
        "description": "The related Payment entities to the Registration.",
        "type": "array",
        "relation": true,
        "relation_type": "EE_HABTM_Relation",
        "readonly": true,
        "joining_model_name": "Registration_Payment",
        "items": {
          "type": "object"
        },
        "relation_model": "Payment"
      },
      "messages": {
        "description": "The related Message entities to the Registration.",
        "type": "array",
        "relation": true,
        "relation_type": "EE_Has_Many_Any_Relation",
        "readonly": true,
        "items": {
          "type": "object"
        },
        "relation_model": "Message"
      },
      "extra_metas": {
        "description": "The related Extra_Meta entities to the Registration.",
        "type": "array",
        "relation": true,
        "relation_type": "EE_Has_Many_Any_Relation",
        "readonly": true,
        "items": {
          "type": "object"
        },
        "relation_model": "Extra_Meta"
      },
      "change_logs": {
        "description": "The related Change_Log entities to the Registration.",
        "type": "array",
        "relation": true,
        "relation_type": "EE_Has_Many_Any_Relation",
        "readonly": true,
        "items": {
          "type": "object"
        },
        "relation_model": "Change_Log"
      }
```

**Comment by nerrad** [comment]
more about schema requests here: https://github.com/eventespresso/event-espresso-core/blob/master/docs/C--REST-API/ee4-rest-api-schema.md

**Comment by jasonnehf** [comment]
Thanks to you both! 

[rolls up sleeves]

Time to get back to it! :)

**Comment by joshfeck** [comment]
Closing this due to inactivity.

**Comment by fogunkoya** [comment]
Hi @jasonnehf did you make any progress with this. Trying to do the same and its proving quite cumbersome

**Comment by jasonnehf** [comment]
Not really.  🙁 

As I'm sure you're discovering, each new registration requires several requests to be made to the API to generate and initialize the (rather large number of) objects that are necessary to represent a successful registration & all of its related structures. What's needed is a walkthrough/example that lists exactly what those related structures are and how they should be initialized to ensure that the final gestalt registration is functionally equivalent to one made via the website. 

**Without such documentation, API-born registrations are doomed to be broken**.

I found that, even after all the _strictly_-required pieces were created and an API-born registration was created, there were still aspects of those registrations that didn't work properly. A piece here, some glue there... More often than not, these registrations would return 500 Server Error messages when interacting with them via the WP Admin panel — something our client was very vocal about.

Frankly, **I didn't have the time or the patience** to scour the numerous models & controllers **to find out how to properly initialize the missing structures**, and I got the impression that **the developers didn't have the time or the incentive to provide any real-world, start-to-finish examples** of how to *actually use* the new feature they worked so hard to provide.

Ultimately, the attempt was abandoned — we couldn't create registrations via the API that behaved 100% the way "regular" registrations do, making them unreliable and therefore useless to the client.

I wish I had better news for you. 🙁
